eessi/testsuite/tests/apps/PyTorch/PyTorch_torchvision.py

```diff
@@ -0,0 +1,153 @@
+    scale = parameter(SCALES.keys())
+    # Not sure how we would ensure the horovod module is _also_ loaded...
+    # parallel_strategy = parameter([None, 'horovod', 'ddp'])
+    parallel_strategy = parameter([None, 'ddp'])
+    compute_device = variable(str)
+    # module_name = parameter(find_modules('PyTorch-bundle'))
+    module_name = parameter(find_modules('torchvision'))
+
+    descr = 'Benchmark that runs a selected torchvision model on synthetic data'
+
+    executable = 'python'
+
+    valid_prog_environs = ['default']
+    valid_systems = ['*']
+
+    time_limit = '30m'
+
+    @run_after('init')
+    def prepare_test(self):
+
+        # Set nn_model as executable option
+        self.executable_opts = ['pytorch_synthetic_benchmark.py --model %s' % self.nn_model]
+
+        # If not a GPU run, disable CUDA
+        if self.compute_device != DEVICE_TYPES[GPU]:
+            self.executable_opts += ['--no-cuda']
+
+
+
+    @run_after('init')
+    def apply_init_hooks(self):
+        # Filter on which scales are supported by the partitions defined in the ReFrame configuration
+        hooks.filter_supported_scales(self)
+
+        # Make sure that GPU tests run in partitions that support running on a GPU,
+        # and that CPU-only tests run in partitions that support running CPU-only.
+        # Also support setting valid_systems on the cmd line.
+        hooks.filter_valid_systems_by_device_type(self, required_device_type=self.compute_device)
+
+        # Support selecting modules on the cmd line.
+        hooks.set_modules(self)
+
+        # Support selecting scales on the cmd line via tags.
+        hooks.set_tag_scale(self)
+
+    @run_after('init')
+    def set_tag_ci(self):
+        if self.nn_model == 'resnet50':
+             self.tags.add(TAGS['CI'])
+
+    @run_after('setup')
+    def apply_setup_hooks(self):
+        if self.compute_device==DEVICE_TYPES[GPU]:
+            hooks.assign_tasks_per_compute_unit(test=self, compute_unit=COMPUTE_UNIT[GPU])
+        else:
+            # Hybrid code, so launch 1 rank per socket.
+            # Probably, launching 1 task per NUMA domain is even better, but the current hook doesn't support it
+            hooks.assign_tasks_per_compute_unit(test=self, compute_unit=COMPUTE_UNIT[CPU_SOCKET])
+
+        # This is a hybrid test, binding is important for performance
+        hooks.set_compact_process_binding(self)
+
+    @run_after('setup')
+    def set_ddp_env_vars(self):
+        # Set environment variables for PyTorch DDP
+        ### TODO: THIS WILL ONLY WORK WITH SLURM, WE SHOULD MAKE A SKIP_IF BASED ON THE SCHEDULER
+        if self.parallel_strategy == 'ddp':
+            self.prerun_cmds = [
+                'export MASTER_PORT=$(expr 10000 + $(echo -n $SLURM_JOBID | tail -c 4))',
+                'export WORLD_SIZE=%s' % self.num_tasks,
+                'echo "WORLD_SIZE="${WORLD_SIZE}',
+                'master_addr=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)',
```

Comment:
Asked the ReFrame devs if this (getting the node hostnames in the allocation) can be done in ReFrame. If htey can't support it, make a hook out of this. Then we can add support for other launchers by updating the hook.

Similar for exporting environments in a parallel-launcher agnostic way.

Reply:
https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#reframe.core.schedulers.Job.nodelist => this is in ReFrame, but also only supports SLURM. But we could potentially contribute others.